language/move-compiler/src/hlir/translate.rs

```diff
@@ -1323,10 +1363,26 @@ fn exp_impl(
         }
         TE::Borrow(mut_, te, f) => {
             let e = exp(context, result, None, *te);
+            if mut_ {
+                match &e.exp.value {
+                    HE::Move { var, .. } | HE::Copy { var, .. } => {
+                        context.mutably_used_locals.insert(*var);
+                    }
+                    _ => (),
+                }
+            }
             HE::Borrow(mut_, e, f)
         }
         TE::TempBorrow(mut_, te) => {
             let eb = exp_(context, result, None, *te);
+            if mut_ {
+                match &eb.exp.value {
+                    HE::Move { var, .. } | HE::Copy { var, .. } => {
```

Comment:
For all these, I think you can do `if let HE::Move { var, .. } | HE::Copy { var, .. } = &eb.exp.value`

I'm not sure one is more readable than the other though